src/portable/microchip/sam7x/dcd_sam7x.c

```diff
@@ -0,0 +1,710 @@
+// As SAM7x has only 4KB packet buffer, use with caution !
+// Enable in FS mode as packets are smaller
+#ifndef USE_DUAL_BANK
+#  if TUD_OPT_HIGH_SPEED
+#    define USE_DUAL_BANK   0
+#  else
+#    define USE_DUAL_BANK   1
+#  endif
+#endif
+
+#define EP_MAX            10
+
+#define USBHS_RAM_ADDR  0xA0100000u
+
+#define EP_GET_FIFO_PTR(ep, scale) (((TU_XSTRCAT(TU_STRCAT(uint, scale),_t) (*)[0x8000 / ((scale) / 8)])USBHS_RAM_ADDR)[(ep)])
+
+// Errata: The DMA feature is not available for Pipe/Endpoint 7
+#define EP_DMA_SUPPORT(epnum) (epnum >= 1 && epnum <= 6)
+
+// DMA Channel Transfer Descriptor
+typedef struct {
+  volatile uint32_t next_desc;
+  volatile uint32_t buff_addr;
+  volatile uint32_t chnl_ctrl;
+  uint32_t padding;
+} dma_desc_t;
+
+// Transfer control context
+typedef struct {
+  uint8_t * buffer;
+  uint16_t total_len;
+  uint16_t queued_len;
+  uint16_t max_packet_size;
+  uint8_t interval;
+  tu_fifo_t * fifo;
+} xfer_ctl_t;
+
+static tusb_speed_t get_speed(void);
+static void dcd_transmit_packet(xfer_ctl_t * xfer, uint8_t ep_ix);
+
+// DMA descriptors shouldn't be placed in ITCM
+#if defined(USB_DMA_DESC_SECTION)
+TU_ATTR_SECTION(TU_XSTRING(USB_DMA_DESC_SECTION))
+#endif
+dma_desc_t dma_desc[6];
+
+xfer_ctl_t xfer_status[EP_MAX+1];
+
+static const tusb_desc_endpoint_t ep0_desc =
+{
+  .bEndpointAddress = 0x00,
+  .wMaxPacketSize   = { .size = CFG_TUD_ENDPOINT0_SIZE },
+};
+
+//------------------------------------------------------------------
+// Device API
+//------------------------------------------------------------------
+
+// Initialize controller to device mode
+void dcd_init (uint8_t rhport)
+{
+  // Enable USBPLL
+  PMC->CKGR_UCKR = CKGR_UCKR_UPLLEN | CKGR_UCKR_UPLLCOUNT(0x3fU);
+  // Wait until USB UTMI stabilize
+  while (!(PMC->PMC_SR & PMC_SR_LOCKU));
+#if !TUD_OPT_HIGH_SPEED
+  // Enable USB FS clk
+  PMC->PMC_MCKR &= ~PMC_MCKR_UPLLDIV2;
+  PMC->PMC_USB = PMC_USB_USBS | PMC_USB_USBDIV(10 - 1);
+  PMC->PMC_SCER = PMC_SCER_USBCLK;
+#endif
+  dcd_connect(rhport);
+}
+
+// Enable device interrupt
+void dcd_int_enable (uint8_t rhport)
+{
+  (void) rhport;
+  NVIC_EnableIRQ((IRQn_Type) ID_USBHS);
+}
+
+// Disable device interrupt
+void dcd_int_disable (uint8_t rhport)
+{
+  (void) rhport;
+  NVIC_DisableIRQ((IRQn_Type) ID_USBHS);
+}
+
+// Receive Set Address request, mcu port must also include status IN response
+void dcd_set_address (uint8_t rhport, uint8_t dev_addr)
+{
+  (void) dev_addr;
+  // DCD can only set address after status for this request is complete
+  // do it at dcd_edpt0_status_complete()
+
+  // Response with zlp status
+  dcd_edpt_xfer(rhport, tu_edpt_addr(0, TUSB_DIR_IN), NULL, 0);
+}
+
+// Wake up host
+void dcd_remote_wakeup (uint8_t rhport)
+{
+  (void) rhport;
+  USBHS->USBHS_DEVCTRL |= USBHS_DEVCTRL_RMWKUP;
+}
+
+// Connect by enabling internal pull-up resistor on D+/D-
+void dcd_connect(uint8_t rhport)
+{
+  (void) rhport;
+  uint32_t irq_state = __get_PRIMASK();
+  __disable_irq();
```

Comment:
do we really need to disable the whole irq, can we just disable the USB IRQ for this.

Reply:
Maybe, I just copied ASF's code.

Reply:
But ` __disable_irq()` is still needed in https://github.com/hathach/tinyusb/blob/f039607afcf7473ea3869f2e2e7cd763d13f8de0/src/portable/microchip/sam7x/dcd_sam7x.c#L584

Reply:
that is quite a lot, we don't disconnect/connect often, but that is a bit weird to go into critical section for this. But let just leave it as it is. We can revisit it later.